Notifications - issue #165

zupan opened the pull request:
Notifications about new comments to wall posts.

zupan committed:
Done basic logic in javascript and add some lines in css.

zupan committed:
Added some css and new subscribers fiels in Post (api/models.py).

zupan committed:
When post is created author is added to the list of subscribers.

mitar commented on piplmesh/account/models.py:
```diff
@@ -23,6 +23,10 @@ class Connection(mongoengine.EmbeddedDocument):
 
+class Notification(mongoengine.EmbeddedDocument):
+    read = mongoengine.BooleanField()
+    comment = mongoengine.IntField()
```
Int field? What will be stored here?

mitar commented on piplmesh/api/models.py:
```diff
@@ -33,6 +35,7 @@ class Post(base.AuthoredDocument):
     attachments = mongoengine.ListField(mongoengine.EmbeddedDocumentField(Attachment), default=lambda: [], required=False)
 
+    subscribers = mongoengine.ListField(mongoengine.ReferenceField(account_models.User), default=lambda: [], required=False)
     # TODO: Prevent posting comments if post is not published
```
Empty line between subscribers and TODOs.

mitar commented on piplmesh/api/models.py:
```diff
@@ -4,6 +4,8 @@
 
 import mongoengine
 
+from piplmesh.account import models as account_models
```
Import with relative import as you can see bellow.

mitar commented:
Please update from main repository as some things were changed in how JavaScript is included.

mitar commented on src/mongodbforms:
```diff
@@ -0,0 +1 @@
+Subproject commit 01201c7b14127a3cb2cb027965320ed8ec499a4e
```
WHAT IS THIS?

mitar commented on src/mongogeneric:
```diff
@@ -0,0 +1 @@
+Subproject commit 0d9d320398865d04e5c9d72b47543ecf68471813
```
WHAT IS THIS?

mitar commented on src/pip-delete-this-directory.txt:
```diff
@@ -0,0 +1,5 @@
+This file is placed here by pip to indicate the source was put
```
WHAT IS THIS?

mitar commented on piplmesh/api/resources.py:
```diff
@@ -63,9 +72,14 @@ class PostResource(AuthoredResource):
     attachments = fields.EmbeddedListField(of='piplmesh.api.resources.AttachmentResource', attribute='attachments', default=lambda: [], null=True, full=True)
-
+    def obj_create(self, bundle, request=None, **kwargs):
+        obj = super(PostResource, self).obj_create(bundle, request=request, **kwargs)
```
OK, obj_create returns `bundle`, not `obj`. Code would be clearer if you would do it like that.

mitar replied:
Probably better place for all this would be in `hydrate`, not `obj_create`:

http://django-tastypie.readthedocs.org/en/latest/resources.html#hydrate

mitar replied:
See how we are populating fields in `AuthoredResource`.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,6 +26,15 @@ def hydrate(self, bundle):
+        obj = super(CommentResource, self).obj_create(bundle, request=request, **kwargs)
+        
+        print obj.obj.author
+        #self.parent.fields['subscribers'].append(obj.obj.author)
```
`self.instance` is what you search for (post object).

zupan committed:
Merging...

zupan committed:
Made some changes, fixed subscribers list.

zupan committed:
New comments are appending to user's notification list.

mitar commented:
Please update from main repository.

zupan committed:
Updated from main repository. Minor changes in resources.py

zupan commented:
OK, I updated this pull request. I have a question now that entries are successfully added to user's notification list when post or comment is written. Should i write part where notification list is drawn and updated in folder "panels" like onlineusers is?

mitar commented on piplmesh/api/resources.py:
```diff
@@ -3,6 +3,7 @@
 from tastypie_mongoengine import fields, resources
 
 from piplmesh.account import models as account_models
+
```
No need, because they are all part of the same package.

mitar commented on piplmesh/api/models.py:
```diff
@@ -4,6 +4,8 @@
 import mongoengine
 
+from piplmesh.account import models as account_models
+
```
No need, because they are all part of the same package.

mitar commented on piplmesh/frontend/static/piplmesh/js/notifications.js:
```diff
@@ -0,0 +1,57 @@
```
No need for additional file. Just put it into `home.js`.

zupan replied:
Finished script will be placed is home.js.

mitar commented:
> Should i write part where notification list is drawn and updated in folder "panels" like onlineusers is?

No! Notification should be a drop-down list from the header. Like Facebook has. Panels are for content.

mitar commented on piplmesh/account/models.py:
```diff
@@ -32,6 +32,10 @@ class Connection(mongoengine.EmbeddedDocument):
 
+class Notification(mongoengine.EmbeddedDocument):
+    read = mongoengine.BooleanField()
+    comment = mongoengine.StringField()
```
String? No, why not a reference field to Comment document?

But you probably get an circular import error. :-)

mitar replied:
Maybe this shows that list of notifications should not be in the User document. Which is also reasonable. If you think, much better is to have a separate collections of all notifications, with reference to user and comment. And then do a query on it when you want to display notifications. Why would you get a list of ALL (past and current) notifications for the user everytime you access user object?

zupan replied:
No I don't get any errors with StringField. When you post comments entry is added to every user that is on superscription list. I placed notifications in user model because we said so when we were deciding what would be the best solution. But i can change it.

mitar replied:
You would get error, if you would do reference to the Comment document. :-) With string you do not get it, because you do not have to import Comment document. :-)

Yes, I think it would be better to move notifications to separate document.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,6 +27,20 @@ def hydrate(self, bundle):
 class CommentResource(AuthoredResource):
+    def obj_create(self, bundle, request=None, **kwargs):
+        bundle = super(CommentResource, self).obj_create(bundle, request=request, **kwargs)
+        bundle.obj.resource_uri = "/api/v1/post/%s/comments/%s" % (self.instance.id, bundle.obj.pk)
```
O no, no no no. Never create URL manually!

zupan replied:
Can you suggest how should i referenced it otherwise? Comment doesn't have id field.

mitar replied:
Then add id field? :-) Simple.

So also URL is not good, because it is only an index in the list. If we will support deleting comments some day, this will not work anymore.

But here I didn't like that you was doing URL manually. So you hard-coded the URL. What if REST interface would not be under `/api/` anymore? This code would break!

zupan replied:
Ok, will do.

mitar commented:
So, please fix the model. Also, I am missing push updates. You have to push to the clients new notifications. Or will this come later?

This is for now. Ball is yours.

zupan committed:
Changed model's structure and move notifications from user to new model. Add notification to push server.

zupan committed:
Created notification api so that fields in api are the same as fields sent to push server.

zupan commented:
So, i fixed the model and other things that you weren't pleased with. Notifications are now separated from User model. I also added push updates and created notifications api. Fields in notifications api and in push updates are the same so that drawing notifications list will be easy.

zupan commented:
Now i have to write JS part so that notification list will be displaying and updating.

mitar commented:
Yes. And displaying the counter. And selecting the ones which have been read. And a separate view which shows you all notifications/whole activity history.

The tricky part will probably be linking from notifications. So how to link to a comment/post where something happened. You can just provide a stub function for that for now and we will fill this in later, when we will have code for posts and comments and will then scroll the post into the focus.

zupan committed:
Add javascript code which displays simple notification list.

zupan committed:
Fixed resources.py so that resource_uri is also pushed with other notification data and added some css for notification list.

zupan commented:
I have question regarding celerybeat. I followed documentation on url http://django-pushserver.readthedocs.org/en/latest/usage.html but i'm getting an error in terminal where celerybeat is running. Error is about online_users
"entity = self._declare_queue(queue, retry, _retry_policy)
  File "C:\Python27\lib\site-packages\celery\app\amqp.py", line 168, in _declare_queue
    queue = messaging.entry_to_queue(name, **options)(self.channel)
SchedulingError: Couldn't apply scheduled task check_online_users: 'module' object has no attribute 'entry_to_queue'"

This is just last few lines. I cannot finish notifications if i am getting this error. Do you have any idea why is this occurring?

mitar commented:
Probably you have some version mismatch. One version of some dependency newer or older that what some other library requires. Compare your `pip freeze` with the one from `requirements.txt`. Do all versions match?

mitar commented on .gitignore:
```diff
@@ -27,3 +27,7 @@ celerybeat-schedule
+
+# PyDev project file & .metadata file
+.pydevproject
+/.metadata/
```
This could be done simply by adding `.*`, so to ignore all hidden files and directories.

mitar commented on piplmesh/account/models.py:
```diff
@@ -79,6 +79,8 @@ class User(auth.User):
     connection_last_unsubscribe = mongoengine.DateTimeField()
     is_online = mongoengine.BooleanField(default=False)
 
+    #notifications = mongoengine.ListField(mongoengine.EmbeddedDocumentField(Notification), default=lambda: [], required=False)
```
Don't forget to remove this.

mitar commented on piplmesh/api/models.py:
```diff
@@ -42,6 +46,24 @@ def save(self, *args, **kwargs):
 
+class Notification(base.AuthoredDocument):
+    """
+    This class defines document type for notifications
```
Dot at the end.

mitar commented on piplmesh/api/models.py:
```diff
@@ -42,6 +46,24 @@ def save(self, *args, **kwargs):
+
+    read = mongoengine.BooleanField(default=False)
+    post = mongoengine.ReferenceField(Post)
+    comment = mongoengine.IntField()
```
No, int is not good ID. But OK, use this for now, but add TODO for later, maybe somebody else can improve this later.

mitar commented on piplmesh/api/models.py:
```diff
@@ -42,6 +46,24 @@ def save(self, *args, **kwargs):
+    comment = mongoengine.IntField()
+    
+    @classmethod
+    def add_notification(cls, user, post, comment_pk):
```
This is not needed, MongoEngine already has [create](http://mongoengine-odm.readthedocs.org/en/latest/apireference.html#mongoengine.queryset.QuerySet.create) method.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +31,73 @@ def hydrate(self, bundle):
         return bundle
 
 class CommentResource(AuthoredResource):
+    def obj_create(self, bundle, request=None, **kwargs):
```
Please read my comments here:

https://github.com/wlanslovenija/PiplMesh/pull/199/files#r1366928

So the main point is that you should decouple API from frontend. So in the API you just deal with storing data (so storing to list of subscribers is OK). But to send something for frontend (`send_update`) is no go.

So please define and use Django signals. And move the code for frontend to frontend.

mitar replied:
And same as for discussion above, be careful where you put those signals. So adding subscribers is a part of creating an object. But sending notifications should come later, when we are really sure everything was done properly and we are just about to return the response.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +31,73 @@ def hydrate(self, bundle):
+                    },
+                })
+
+        if bundle.obj.author not in self.instance.subscribers:
```
This part can stay here. So when comment is created, author is automatically added to the list of subscribers. This is something which is part of the resource-handling, yes.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -63,9 +129,16 @@ class PostResource(AuthoredResource):
+    
     class Meta:
         queryset = api_models.Post.objects.all()
+        #print queryset
```
This should probably be removed. :-)

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +31,73 @@ def hydrate(self, bundle):
 
+class NotificationResource(AuthoredResource):
+#    post = fields.ReferenceField(to='piplmesh.api.resources.PostResource', attribute='post', null=False, full=False, readonly=True)
+    post = tastypie_fields.CharField(attribute='post', default='', null=False, blank=True)
```
Why readonly? No, you want to be able to programmaticaly manipulate this fields. You probably just do not want that people create that through API, just read. You limit that by limiting HTTP methods allowed just to GET.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +31,73 @@ def hydrate(self, bundle):
         # TODO: Make proper authorization, current implementation is for development use only
         authorization = tastypie_authorization.Authorization()
 
+class NotificationResource(AuthoredResource):
```
Probably you want to use simply `Resource` class, not `AuthoredResource` which defines some its fields.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +31,73 @@ def hydrate(self, bundle):
         authorization = tastypie_authorization.Authorization()
 
+class NotificationResource(AuthoredResource):
+#    post = fields.ReferenceField(to='piplmesh.api.resources.PostResource', attribute='post', null=False, full=False, readonly=True)
```
Why this does not work?

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +31,73 @@ def hydrate(self, bundle):
+class NotificationResource(AuthoredResource):
+#    post = fields.ReferenceField(to='piplmesh.api.resources.PostResource', attribute='post', null=False, full=False, readonly=True)
+    post = tastypie_fields.CharField(attribute='post', default='', null=False, blank=True)
+    author = fields.ReferenceField(to='piplmesh.api.resources.UserResource', attribute='author', null=False, full=True, readonly=True)
```
Not `author`. It is not `author` who has notification.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +31,73 @@ def hydrate(self, bundle):
+#    post = fields.ReferenceField(to='piplmesh.api.resources.PostResource', attribute='post', null=False, full=False, readonly=True)
+    post = tastypie_fields.CharField(attribute='post', default='', null=False, blank=True)
+    author = fields.ReferenceField(to='piplmesh.api.resources.UserResource', attribute='author', null=False, full=True, readonly=True)
+    message = tastypie_fields.CharField(attribute='message', default='', null=False, blank=True)
```
Message? What is this?

mitar commented on piplmesh/frontend/context_processors.py:
```diff
@@ -11,7 +11,7 @@ def global_vars(request):
     return {
         # Constants
-        'HOME_CHANNEL_ID': views.HOME_CHANNEL_ID,
+        'HOME_CHANNEL_ID': settings.HOME_CHANNEL_ID,
```
Why move to settings? This is a program constant. Not an user setting.

mitar commented on piplmesh/frontend/static/piplmesh/css/style.css:
```diff
@@ -154,6 +154,59 @@ form input[type="submit"], form input[type="reset"] {
+
+li.notification {
+    border-bottom: 1px solid #68AE2C;
+    text-align: left !important;
```
Are you sure you cannot solve this without `!important` flag? As we have all CSS in our hands, this is strange to be needed.

mitar commented on piplmesh/frontend/templates/header.html:
```diff
@@ -32,6 +32,19 @@
                 <li class="username"><a href="{% url "profile" username=user.username %}">{{ user.username }}</a></li>
                 {% if user.is_authenticated %}
+                    <li>
+                        <div class="notifications">
```
Why only if authenticated? All users can get notifications, no? We are using lazy users now, remember. authenticated means that user is registered. But all users can post content.

mitar commented on piplmesh/frontend/templates/home.html:
```diff
@@ -7,6 +7,7 @@
 {% block js %}
     {{ block.super }}
     <script src="{% static "pushserver/updates.js" %}" type="text/javascript"></script>
+    <script src="{% static "piplmesh/js/notifications.js" %}" type="text/javascript"></script>
```
No need for this to in a separate file? We will not have notifications on other pages? Or will we? Hm, maybe. :-)

mitar commented:
> I have question regarding celerybeat.

I just pushed new `requirements.txt` which is working for me. Verified. Please pull from main repository and upgrade to packages from that file.

Commit:
Merge upstream...

Commit:
Merge remote-tracking branch 'upstream/master' into notifications_#165

Commit:
Fixed css and html for notifications, changed fields in Notification and how push for notifications is handled.

zupan commented:
I fixed almost everything that you suggested. But there is one problem. In moved send_update for resources.py. I am using mongoengine signals but there is a problem that i can't figure it out. Signal is working if i print something but when i place update.sent_update i get error 500 from server.

zupan commented:
Also i have a question regarding messages. After i updated to last version signals and messages didn't work. I found out that 'django.contrib.messages.middleware.MessageMiddleware', was missing from settings and that signals.py weren't used anywhere in project. Was this done intentionally?

You said that i have to use signals for sent_update (i did as you can see in previous comment) but where should i place this code. I know that this is frontend, but is it ok to be in models or should i created signals.py (like for django signals).

mitar commented:
I am warning you that new commits were made with some other commiter data. This could lead to complications when I will be grading you.

mitar commented on .gitignore:
```diff
@@ -24,7 +24,7 @@ Thumbs.db
-# Celerybeat
-*.pid
+# Ignore all hidden files and directories
+.*
```
But this is already in the file. :-) At the top.

mitar commented on .gitignore:
```diff
@@ -24,7 +24,7 @@ Thumbs.db
-*.pid
+# Ignore all hidden files and directories
+.*
+/.*
```
Not needed.

mitar commented on piplmesh/account/models.py:
```diff
@@ -1,6 +1,6 @@
 from __future__ import absolute_import
 
-import datetime, hashlib, urllib
+import datetime, hashlib, urllib, os
```
Order alphabetically.

mitar commented on piplmesh/account/models.py:
```diff
@@ -12,6 +12,7 @@
 from django.utils.translation import ugettext_lazy as _
 
 import mongoengine
+
```
No empty line. This is the same package.

mitar commented on piplmesh/account/models.py:
```diff
@@ -62,6 +63,7 @@ class User(auth.User):
     birthdate = fields.LimitedDateTimeField(upper_limit=upper_birthdate_limit, lower_limit=lower_birthdate_limit)
     gender = fields.GenderField()
     language = fields.LanguageField()
+    channel_id = mongoengine.StringField()
```
Maybe just use `UUIDField`?

mitar commented on piplmesh/account/signals.py:
```diff
@@ -7,11 +7,20 @@ def user_login_message(sender, request, user, **kwargs):
+    if not user.channel_id:
+        user.generate_channel_id()
+
+signals.user_logged_in.connect(user_generate_channel_id, dispatch_uid=__name__ + '.user_generate_channel_id')
```
No, we said that we will regenerate it on user disconnecting from the website.

mitar replied:
And login != connecting. Login is when user authenticates. Connecting is when it establishes HTTP PUSH channel. We want to cycle in the later case. Login can be a very rare event.

mitar replied:
Also, we want to create a new one, so even if old one exists, user should get a new channel id.

mitar commented on piplmesh/api/models.py:
```diff
@@ -4,6 +4,10 @@
+from pushserver.utils import updates
+
+from piplmesh.account import models as account_models
+
```
No empty line here. Those two are same package (PiplMesh).

mitar replied:
You can use relative import also for above.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -1,8 +1,13 @@
+from pushserver.utils import updates
+
 from piplmesh.account import models as account_models
+
```
No empty line.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +31,53 @@ def hydrate(self, bundle):
 class CommentResource(AuthoredResource):
+    def obj_create(self, bundle, request=None, **kwargs):
+        bundle = super(CommentResource, self).obj_create(bundle, request=request, **kwargs)
+
```
So, still, you should define a new signal here and then hook onto it.

mitar replied:
Forget it. It is OK like this. We can define notifications as part of the API. Also expose notifications through RESTful API.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +31,53 @@ def hydrate(self, bundle):
+    def get_object_list(self, request):
+        obj_list = super(NotificationResource, self).get_object_list(request)
+        if request:
+            obj_list = obj_list.filter(recipient=request.user)
```
This is not correct way to do it. Tastypie supports [authorization and authentication](http://django-tastypie.readthedocs.org/en/latest/authentication_authorization.html). You should use that. So define a authorization class allows access to only user's notifications.

mitar replied:
See example here:

https://github.com/wlanslovenija/PiplMesh/blob/master/piplmesh/api/authorization.py

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +31,53 @@ def hydrate(self, bundle):
+        return bundle
+
+    def dehydrate(self, bundle):
+        bundle.data['recipient'] = bundle.obj.post.comments[bundle.obj.comment].author
```
What exactly are you doing here?

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +31,53 @@ def hydrate(self, bundle):
+
+    class Meta:
+        queryset = api_models.Notification.objects.all()
+        fields = ['id', 'created_time', 'comment', 'resource_uri', 'read', 'post']
```
I am not sure you need to define this?

mitar commented on piplmesh/frontend/context_processors.py:
```diff
@@ -12,6 +12,7 @@ def global_vars(request):
     return {
         # Constants
         'HOME_CHANNEL_ID': views.HOME_CHANNEL_ID,
+        'NOTIFICATION_CHANNEL_ID': 'user/%s' % request.user.channel_id,
```
YES!!!

Rename to `USER_CHANNEL_ID`. It will not be used only for notifications.

mitar commented on piplmesh/frontend/static/piplmesh/js/notifications.js:
```diff
@@ -0,0 +1,105 @@
+
+$(document).ready(function() {
```
Space before `()` in anonymous functions.

mitar commented on piplmesh/frontend/static/piplmesh/js/notifications.js:
```diff
@@ -0,0 +1,105 @@
+
+$(document).ready(function() {
+    $(".notifications").click(function (){
```
Space after `()`. Use `'` for constant strings.

mitar commented on piplmesh/frontend/static/piplmesh/js/notifications.js:
```diff
@@ -0,0 +1,105 @@
+    $(".notifications").click(function (){
+        if ($("#notif_box").hasClass('show')) {
+            $("#notif_box").slideUp('fast');
+            $("#notif_box").toggleClass('show');
```
They beautify of jQuery is that you can chain this together:

```
$('#notif_box').slideUp('fast').toggleClass('show');
```

But I am not sure why you are setting and clearing this class?

mitar commented on piplmesh/frontend/static/piplmesh/js/notifications.js:
```diff
@@ -0,0 +1,105 @@
+    });
+    
+    
+    $.updates.registerProcessor('notification_channel', 'notifications', AddNewNotification);
```
`user_channel`

mitar commented on piplmesh/frontend/static/piplmesh/js/notifications.js:
```diff
@@ -0,0 +1,105 @@
+}
+
+function loadNotifications() {
+    $.getJSON('/api/v1/notification/', function(notifications) {
```
You should not use hard-coded URLs.

mitar replied:
Space before `(notifications)`.

aljana commented:
> I am warning you that new commits were made with some other commiter data. This could lead to complications when I will be grading you.

I had the same problem. I was working on Ubuntu and in one commit suddenly appeared my full name. I have no idea why this happend??

mitar commented:
Yes, I know you had the same problem. I wrote also to your pull request about that.

How you have this configured? Do you have git configuration file for your user? Otherwise it tries to get this from current logged in user.

mitar commented on piplmesh/frontend/static/piplmesh/js/notifications.js:
```diff
@@ -0,0 +1,105 @@
+function formatDate(time) {
+    //var months = new Array("January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December");
+    var days = new Array("Pon", "Tor", "Sre", "Čet", "Pet", "Sob", "Ned");
+    var months = new Array("Jan", "Feb", "Mar", "Apr", "Maj", "Jun", "Jul", "Avg", "Sep", "Okt", "Nov", "December");
```
Those should be market as translatable JavaScript strings. And be in English.

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -24,7 +24,7 @@ def check_online_users():
         ).update(set__is_online=True):
             updates.send_update(
-                views.HOME_CHANNEL_ID,
+                settings.HOME_CHANNEL_ID,
```
Why settings?

mitar commented on piplmesh/frontend/templates/header.html:
```diff
@@ -31,6 +31,19 @@
                     {% user_image %}
                 </li>
                 <li class="username"><a href="{% url "profile" username=user.username %}">{{ user.username }}</a></li>
+                 <li>
```
Indentation.

mitar replied:
TODO.

mitar commented on piplmesh/frontend/templates/home.html:
```diff
@@ -7,11 +7,13 @@
         $.updates.subscribe({
-            'home_channel': '{% filter escapejs %}{% channel_url HOME_CHANNEL_ID %}{% endfilter %}'
+            'home_channel': '{% filter escapejs %}{% channel_url HOME_CHANNEL_ID %}{% endfilter %}',
+            'notification_channel': '{% filter escapejs %}{% channel_url NOTIFICATION_CHANNEL_ID %}{% endfilter %}'
```
Name.

mitar commented on piplmesh/settings.py:
```diff
@@ -150,6 +150,7 @@
 MIDDLEWARE_CLASSES = (
     'django.middleware.common.CommonMiddleware',
     'django.contrib.sessions.middleware.SessionMiddleware',
+    'django.contrib.messages.middleware.MessageMiddleware',
```
This is already there???

mitar commented on piplmesh/urls.py:
```diff
@@ -4,7 +4,7 @@
 from tastypie import api
 
-from piplmesh.account import models, views as account_views
+from piplmesh.account import models, views as account_views, signals
```
Why importing signals?

mitar replied:
TODO.

mitar replied:
TODO.

mitar commented:
> Signal is working if i print something but when i place update.sent_update i get error 500 from server.

Where you get 500? For which URL?

mitar commented:
> I found out that 'django.contrib.messages.middleware.MessageMiddleware', was missing from settings

This is not true? I can see it in settings?

> signals.py weren't used anywhere in project

OK. This is a bug, it seems. In fact, it is wrong to have this in `signals.py`. `signals.py` should be a place where we define our own signals. And then in `views.py` you should register to them. So please move content from `signals.py` to `views.py`.

> You said that i have to use signals for sent_update (i did as you can see in previous comment) but where should i place this code.

For your own signals you define (if you need to, I changed my mind and I think you do not need to anymore). you define it `signals.py`. Implementation of signal receivers you often do in `views.py` (or `models.py`, but I prefer views, but it depends for what signals are, for login/logout views seems more reasonable).

> I know that this is frontend, but is it ok to be in models or should i created signals.py

In fact I don't agree with this anymore. Notifications are part of API. So list of notifications available are part of the API.

The web interface to display them, pushing to this web interface, this is frontend. So this part should be in frontend. Because in theory user could have some other client, like mobile phone. And notifications could be read differently there. And pushed to.

> (like for django signals).

Those are Django signals! :-) Just not bundled ones.

mitar commented:
Ping.

mitar commented:
Please update from main repository.

zupan commented:
I hadn't been active lately because i am preoccupied with studying.

zupan committed:
Changed channel_id, fixed bugs in notification.js, resources.py, models.py.

zupan committed:
Upstream merge.

zupan commented:
I have updated to main repository, but site is full of errors. Registration is not working "can't compare offset-naive and offset-aware datetimes", common.js is alerting every 5 seconds (ajax error). Why is this even in main repository, was this even tested before committing?

mitar commented:
Have you upgraded your modules as defined in `requirements.txt`? ;-)

mitar commented:
But I did just pushed some small bugfixes so you are welcome to pull again.

mitar commented:
So, you should pull, merge, drop the database, and upgrade Python modules. If it still does not work afterwards, please fill bug report.

zupan committed:
Fixed harcoded notification api url.

zupan commented:
I did everything as you wrote, and updated again after you pushed small bugfixes. Problem is the same. Registration had to be broken with some pull request. And why bug reports if they're random annoying debug alerts in javascript that are still in code?

mitar commented:
Are you running push server in the background?

zupan commented:
Yes. I get "TypeError at /register/ can't compare offset-naive and offset-aware datetimes" (account/fields.py:38). And there are javascript alerts in common.js)

mitar commented:
OK. I can confirm the register bug.

I don't understand the alerts. Can you explain more? So alerts are if Ajax requests fail. If they are failing regularly, this means HTTP push code has problems connecting to the push server.

zupan commented:
Reason for alerts is mine, because i am working with api, but are they supposed to be in master (common.js line 8)?

mitar commented:
So you are right if you want to say that this should not be in production like that. But yes, some global error handling for Ajax should be defined in common.js. Just probably it should behave a bit nicer. But for now, for development, is good enough. But if you want, you can fix that (in a separate pull request).

mitar commented:
Fixed registration problem.

zupan committed:
Fixed push for notifications using signals, changing channel_id on disconnect and other minor issues.

zupan committed:
Merge remote-tracking branch 'upstream/master' into notifications_#165

zupan commented:
Fixed almost everything except for javascript in notifications.js. This will come in next commit.

mitar commented on .gitignore:
```diff
@@ -24,7 +24,3 @@ Thumbs.db
-*.celerybeat-schedule
-
-# Celerybeat
-*.pid
```
Any particular reason that you are removing those lines? For different platforms we have observed different files so this is why all this is here. And I don't think there are minuses to having this here. Why then removing?

mitar commented on build/pip-delete-this-directory.txt:
```diff
@@ -0,0 +1,5 @@
```
But this should definitely not be committed!

mitar commented on piplmesh/account/models.py:
```diff
@@ -27,6 +27,9 @@ def upper_birthdate_limit():
     return timezone.now().date() - datetime.timedelta(LOWER_DATE_LIMIT)
 
+def generate_channel_id():
+    return os.urandom(16).encode('hex')
```
I don't like hex encoding. It has low entropy (only 16 characters are used) in comparison with the whole alphanumeric range. So if you would use whole alphanumeric range, you should encode the same random number much shorter. Thus have much shorter channel_id and URLs, but same randomness. Use `django.utils.crypto.get_random_string`.

But it is good to have a separate function for this, so that later on you can just change this function if the way we want to generate those IDs should be different.

mitar commented on piplmesh/account/models.py:
```diff
@@ -62,6 +65,7 @@ class User(auth.User):
     birthdate = fields.LimitedDateTimeField(upper_limit=upper_birthdate_limit, lower_limit=lower_birthdate_limit)
     gender = fields.GenderField()
     language = fields.LanguageField()
+    channel_id = mongoengine.UUIDField(binary=False, default=generate_channel_id)
```
Oh, you are using UUIDField. Then you should be using uuid value as a default. You can use random uuid for this. (See uuid Python package.)

zupan replied:
Fixed.

mitar commented on piplmesh/account/models.py:
```diff
@@ -155,6 +159,9 @@ def get_image_url(self):
             return staticfiles_storage.url(settings.DEFAULT_USER_IMAGE)
 
+    def get_user_channel(self):
+        return "user_%s" % self.channel_id
```
You can use `user/%s`.

zupan replied:
No i can't. In settings.py there is a regex  code "url': r'/updates/([^/]+)/", where "/" is excluded. I tried using / but this was the reason for error 500 that i mentioned some time ago. Then I changed regex code so that "/" is allowed but this opens whole new list of errors. I think there is a reason for  excluded "/" in push server url.

mitar replied:
There is no reason for `/` be excluded. I am using in some other projects the following configuration and it seems to work nicely:

```
# URL prefix for django-pushserver passthrough callbacks
PUSH_SERVER_URL = '/passthrough/'

PUSH_SERVER = {
    'port': 8001,
    'address': '127.0.0.1',
    'store': {
        'type': 'memory',
        'min_messages': 0,
        'max_messages': 100,
        'message_timeout': 10,
    },
    'locations': (
        {
            'type': 'subscriber',
            'url': r'/updates/(.+)/',
            'polling': 'long',
            'create_on_get': True,
            'allow_origin': 'http://127.0.0.1:8000',
            'allow_credentials': True,
            'passthrough': 'http://127.0.0.1:8000' + PUSH_SERVER_URL,
        },
        {
            'type': 'publisher',
            'url': r'/send-update/(.+)/',
        },
    ),
}
```

zupan replied:
Ok. I got it working.

mitar replied:
Great!

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +30,52 @@ def hydrate(self, bundle):
+
+    def obj_create(self, bundle, request=None, **kwargs):
+        bundle = super(NotificationResource, self).obj_create(bundle, request=request, **kwargs)
+        print "Res: :)"
```
What is this?

mitar commented on piplmesh/frontend/static/piplmesh/js/common.js:
```diff
@@ -6,7 +6,7 @@ $(document).ready(function () {
     $(document).ajaxError(function (event, jqXHR, ajaxSettings, thrownError) {
         window.console.error(event, jqXHR, ajaxSettings, thrownError);
-        alert(gettext("Oops, something went wrong..."));
+        //alert(gettext("Oops, something went wrong..."));
```
Don't just comment it out without adding a TODO comment or some other comment with a reason.

mitar commented on piplmesh/settings.py:
```diff
@@ -160,6 +160,7 @@
     'django.middleware.csrf.CsrfViewMiddleware',
     'django.contrib.auth.middleware.AuthenticationMiddleware',
     'piplmesh.account.middleware.LazyUserMiddleware',
+    'django.contrib.messages.middleware.MessageMiddleware',
```
This middleware is already there!

zupan replied:
Removed

mitar replied:
You removed the wrong one!!!

mitar commented on piplmesh/frontend/context_processors.py:
```diff
@@ -13,6 +13,7 @@ def global_vars(request):
     return {
         # Constants
         'HOME_CHANNEL_ID': views.HOME_CHANNEL_ID,
+        'USER_CHANNEL_ID': request.user.get_user_channel(),
```
See, user channel is not a constant but is something you get from user in request. No need to have this in context processor.

mitar commented on piplmesh/frontend/templates/home.html:
```diff
@@ -8,17 +8,20 @@
         $.updates.subscribe({
-            'home_channel': '{% filter escapejs %}{% channel_url HOME_CHANNEL_ID %}{% endfilter %}'
+            'home_channel': '{% filter escapejs %}{% channel_url HOME_CHANNEL_ID %}{% endfilter %}',
+            'user_channel': '{% filter escapejs %}{% channel_url USER_CHANNEL_ID %}{% endfilter %}'
```
Instead of `USER_CHANNEL_ID` you could simply use `user.get_user_channel` here, no?

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +30,52 @@ def hydrate(self, bundle):
         # TODO: Make proper authorization, current implementation is for development use only
         authorization = tastypie_authorization.Authorization()
 
+class NotificationAuthorization(tastypie_authorization.Authorization):
```
Move this to `authorization.py` file.

mitar commented on piplmesh/api/signals.py:
```diff
@@ -1,3 +1,4 @@
 from django import dispatch
 
 post_created = dispatch.Signal(providing_args=('post', 'request', 'bundle'))
+notification_created = dispatch.Signal(providing_args=('notification', 'request'))
```
Add also bundle. You never know what is useful to others.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +30,52 @@ def hydrate(self, bundle):
+    def obj_create(self, bundle, request=None, **kwargs):
+        bundle = super(NotificationResource, self).obj_create(bundle, request=request, **kwargs)
+        print "Res: :)"
+        # signals.notification_created.send(sender=self, notification=bundle.obj, request=request or bundle.request)
```
Why is this commented?

zupan replied:
I forgot to delete this method.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +30,52 @@ def hydrate(self, bundle):
+        bundle.data['content'] = bundle.obj.post.comments[bundle.obj.comment].message
+        return bundle
+
+    def obj_create(self, bundle, request=None, **kwargs):
```
See, there is a difference between REST resource and database object. Here, this is method, which is called, when REST resource wants to create a new database object for it. So it is not called when you create object directly from the code.

mitar replied:
In general, you don't want for users to be able to create notifications over REST. You should disallow this.

zupan replied:
Ok. I wanted to use obj_create to send signal to push notification but signal wasn't trigger-ed (you explained why). I forgot to delete this method before commiting.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +30,52 @@ def hydrate(self, bundle):
+            if subscriber != bundle.obj.author:
+                # add notification to db
+                notification = api_models.Notification.objects.create(recipient=subscriber, post=self.instance, comment=bundle.obj.pk)
+                signals.notification_created.send(sender=self, notification=notification, request=request or bundle.request)
```
MongoEngine already defines [signals for documents](http://mongoengine-odm.readthedocs.org/en/latest/guide/signals.html). Maybe this is what you are searching for here?

zupan replied:
I tried to use mongoengine signals and Post_save method in api.models.py, but there is a problem because i can't import NotificationResources in models.py because of circular imports. Because of that i cannot use serialize from NotificationResources and have to manually create json object using python json package.

This new code is working but is not so clean as current one.
Do i really have to change it?

mitar replied:
But why you want to implement code in models.py? Why not views.py of frontend? The same place where the previous code was? That's the ideas of signals, you can implement them anywhere.

But the question is if that defined signals are good enough. I am not so sure because they are missing resource and bundle. So probably we have to implement our own signals for this case.

mitar replied:
(As you did.)

mitar replied:
(But you have to understand why is necessary to have your own signal. Not because of the location. But because of missing other objects.)

zupan replied:
Yes i understand. I noticed that when using mongoengine signals we are missing objects i just didn't mention it in the comment. That was main reason why i asked if we really wanted to use mongoengine signal.

So should i use django signals instead as i did before because there were located in frontend/views.py and for most important part we had all the object needed?

mitar replied:
> That was main reason why i asked if we really wanted to use mongoengine signal.

And I wanted that you argument why they might not be good. :-)

> So should i use django signals instead

Both MongoEngine and ours are Django signals. :-)

> because there were located in frontend/views.py

This is completely your decision where you decide to connect onto them. There is no reason why you did it in models.py! See, MongoEngine signals are nothing different to signals we do, they are just defined in MongoEngine and called there. Where you connect to them, this is your decision.

> most important part we had all the object needed

This is the most important part. It allows us to same serialization as we are using for REST protocol.

zupan replied:
> And I wanted that you argument why they might not be good. :-)

I did :).

> allows us to same serialization as we are using for REST protocol

I also pointed this out in my first comment when i changed to signals defined in mongoengine.

Signals are now fixed, serialized as we are using for REST protocol and have all the objects that might need in the future.

zupan committed:
Fixes user_chanell_id, mongoengine signals, javascript notification output, notification counter, moved code from notifications.js to home.js.

zupan committed:
Fixed some minor issues, added translation for notifications.

zupan committed:
Removed django.po~.

zupan commented:
Is there anything else that needs to be done in this pull request?

mitar commented on piplmesh/account/models.py:
```diff
@@ -27,6 +29,9 @@ def upper_birthdate_limit():
     return timezone.now().date() - datetime.timedelta(LOWER_DATE_LIMIT)
 
+def generate_channel_id():
+    return uuid.uuid1()
```
Please use `uuid4`.

zupan replied:
Fixed

mitar commented on piplmesh/api/models.py:
```diff
@@ -2,8 +2,11 @@
+import json
 import mongoengine
 
+from pushserver.utils import updates
```
Properly structure imports (each package into separate section).

zupan replied:
Fixed

zupan committed:
Fixed imports, settings.py and generate_channel_id in models.py.

zupan committed:
Fixed signals for pushing notifications.

zupan commented:
Fixed signals.

zupan commented:
I am working on page that shows all notification. I wanted to use ListView from mongogeneric but there is a problem with code "class NotificationsView(detail.ListView):" that i cannot find the solution to:

> 'module' object has no attribute 'ListView'

I am writing this in frontend/views.py. Can you try to import ListView so that we can find an error or solution to this problem. I checked [readme](https://github.com/jschrewe/django-mongogeneric) and it says that ListView is implemented...
Any ideas?

mitar commented on piplmesh/api/authorization.py:
```diff
@@ -8,4 +8,10 @@ def apply_limits(self, request, object_list):
+
+class NotificationAuthorization(tastypie_authorization.Authorization):
+    def apply_limits(self, request, object_list):
+        if request:
```
`and hasattr(request, 'user'):`

mitar commented on piplmesh/api/authorization.py:
```diff
@@ -8,4 +8,10 @@ def apply_limits(self, request, object_list):
+    def apply_limits(self, request, object_list):
+        if request:
+            object_list = object_list.filter(recipient=request.user)
         return object_list
```
Not good. Otherwise it should return an empty list. So you return filtered query set if user information is available, otherwise you return an empty list.

mitar commented on piplmesh/api/models.py:
```diff
@@ -42,6 +46,19 @@ def save(self, *args, **kwargs):
+    """
+
+    recipient = mongoengine.ReferenceField(account_models.User, required=True)
+    created_time = mongoengine.DateTimeField(default=lambda: timezone.now(), required=True)
```
`lambda: timezone.now()` -> `timezone.now`?

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +30,38 @@ def hydrate(self, bundle):
+
+        for subscriber in self.instance.subscribers:
+            if subscriber != bundle.obj.author:
+                # add notification to db
```
This is obvious what you are doing here. Or at least it should be obvious from the code. So `api_models.Notification.objects.create` tells exactly the same as your comment.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +30,38 @@ def hydrate(self, bundle):
+        if bundle.obj.author not in self.instance.subscribers:
+            self.instance.subscribers.append(bundle.obj.author)
+            self.instance.save()
+        return bundle
```
One empty line before this.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +30,38 @@ def hydrate(self, bundle):
+
+    def dehydrate(self, bundle):
+        bundle.data['author'] = bundle.obj.post.comments[bundle.obj.comment].author
+        bundle.data['content'] = bundle.obj.post.comments[bundle.obj.comment].message
```
I don't understand what are you doing here?

mitar replied:
OK. I understand. Because you are storing comment as index. OK. So, there are some problems here. First, you are not calling parent dehydrate!

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +30,38 @@ def hydrate(self, bundle):
         authorization = tastypie_authorization.Authorization()
 
+class NotificationResource(resources.MongoEngineResource):
+    content = tastypie_fields.CharField(attribute='content', default='', null=False, blank=True)
```
Rename this to `comment_message`. There is no need for `attribute` because this is not attribute on the document! So you specify attribute if you want field to be mapped to document field. And you are not doing that. You are populating it manually.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +30,38 @@ def hydrate(self, bundle):
+    content = tastypie_fields.CharField(attribute='content', default='', null=False, blank=True)
+
+    def dehydrate(self, bundle):
+        bundle.data['author'] = bundle.obj.post.comments[bundle.obj.comment].author
```
Where is this field defined in resource? You should define it in a similar manner as you have defined `content` above!

mitar replied:
Rename to `comment_author`.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +30,38 @@ def hydrate(self, bundle):
+class NotificationResource(resources.MongoEngineResource):
+    content = tastypie_fields.CharField(attribute='content', default='', null=False, blank=True)
+
+    def dehydrate(self, bundle):
```
Have you tried to use `dehydrate_<filedname>` approach? Could be cleaner than this.

mitar commented:
Do you anywhere set which notifications are read and which are not?

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
+function addComment(comment) {
+    $.ajax({
+        type: 'POST',
+        url: '/api/v1/post/5050512b6c20b1028c45cf86/scomments/',
```
`scomments`?

mitar replied:
Hardcoded URL? Please build it pragmatically. (You can for example add comment to newest displayed post.)

zupan replied:
This is just test code for creating comments so i can test notification.

mitar replied:
I know. But it is still in pull request and is on the way to the repository. It is nothing wrong if it stays in (with TODO), but it should work for anybody who tries it. And then others can build upon that commenting interface.

mitar commented on piplmesh/frontend/templates/home.html:
```diff
@@ -12,13 +12,15 @@
             'panels_order': '{% filter escapejs %}{% url "panels_order" %}{% endfilter %}',
-            'post': '{% filter escapejs %}{% url "api_dispatch_list" api_name=API_NAME resource_name="post" %}{% endfilter %}'
+            'post': '{% filter escapejs %}{% url "api_dispatch_list" api_name=API_NAME resource_name="post" %}{% endfilter %}',
+            'notifications' : '{% filter escapejs %}{% url "api_dispatch_list" api_name=API_NAME resource_name="notification" %}{% endfilter %}'
```
Unnecessary space.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +30,38 @@ def hydrate(self, bundle):
+        bundle.data['content'] = bundle.obj.post.comments[bundle.obj.comment].message
+        return bundle
+
+    class Meta:
```
You are doing only `GET` on this resource. Maybe you should limit only that GET should be allowed?

zupan replied:
I will also add `PATCH` for updating fields.

mitar replied:
Yes, if you will be also setting read field. But of course, API should allow only setting read field. You might to configure other fields read-only.

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -46,6 +46,8 @@ def check_online_users():
             connections__in=([], None), # None if field is missing altogether
             connection_last_unsubscribe__lt=timezone.now() - datetime.timedelta(seconds=CHECK_ONLINE_USERS_RECONNECT_TIMEOUT),
         ).update(set__is_online=False):
+            user.channel_id = models.generate_channel_id()
```
You should do `user.reload()` before this line because `user` object is obsolete as you have just updated it with `update()`.

mitar commented on piplmesh/frontend/static/piplmesh/css/style.css:
```diff
@@ -211,6 +211,59 @@ input.gsc-search-button:hover,
+    width: 262px;
+}
+
+li.notification:hover{
```
Space before `{`.

mitar commented on piplmesh/frontend/views.py:
```diff
@@ -118,6 +118,24 @@ def send_update_on_new_post(sender, post, request, bundle, **kwargs):
         updates.send_update(HOME_CHANNEL_ID, serialized, True)
 
+@dispatch.receiver(signals.notification_created)
+def send_update_on_new_notification(sender, notification, request, **kwargs):
```
`bundle` argument is missing.

mitar commented on piplmesh/frontend/views.py:
```diff
@@ -118,6 +118,24 @@ def send_update_on_new_post(sender, post, request, bundle, **kwargs):
+    """
+    Sends update to push server when a new notification is created.
+    """
+    serialized = sender.serialize(None, {
```
`None` -> `request`.

mitar commented on piplmesh/frontend/views.py:
```diff
@@ -118,6 +118,24 @@ def send_update_on_new_post(sender, post, request, bundle, **kwargs):
+    Sends update to push server when a new notification is created.
+    """
+    serialized = sender.serialize(None, {
+        'type': 'notifications',
```
`notifications` -> `notification`.

mitar commented on piplmesh/frontend/views.py:
```diff
@@ -118,6 +118,24 @@ def send_update_on_new_post(sender, post, request, bundle, **kwargs):
+    """
+    serialized = sender.serialize(None, {
+        'type': 'notifications',
+        'notifications': {'author' : notification.post.comments[int(notification.comment)].author,
```
`notifications` -> `notification`.

mitar replied:
You should not be required to manually create this dict! This should be done by Tastypie! Check how it is done in `send_update_on_new_post`! And do the same. If there are problems, ask.

mitar replied:
And please wrap things properly.

zupan replied:
How should i do this using Tastypie? I don't get data in bundle like post does. In bundle are data from comment not from notification.

mitar commented on piplmesh/frontend/templates/header.html:
```diff
@@ -31,6 +31,17 @@
                 <li class="username"><a href="{% url "profile" username=user.username %}">{{ user.username }}</a></li>
+                 <li>
+                    <div class="notifications">
+                    </div>
```
No need for new line. You can have both `<div>` and `</div>` on the same line.

mitar commented on piplmesh/frontend/templates/header.html:
```diff
@@ -31,6 +31,17 @@
+                    <div class="notifications">
+                    </div>
+                    <div id="notif_box" class="hide">
+                        <div id="notif_content">
```
No need for new line. You can have both `<div>` and `</div>` on the same line.

mitar replied:
`notif_content` -> be verbose, write out fully.

mitar commented on piplmesh/frontend/templates/header.html:
```diff
@@ -31,6 +31,17 @@
+                    <div id="notif_box" class="hide">
+                        <div id="notif_content">
+                        </div>
+                        <div class='close_notif_box'>
```
Be verbose, write out fully.

mitar commented on piplmesh/frontend/templates/header.html:
```diff
@@ -31,6 +31,17 @@
+                        <div id="notif_content">
+                        </div>
+                        <div class='close_notif_box'>
+                            <div id="addCom">Add c</div> Close
```
Be verbose, write out fully.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
     });
 }
 
+function AddNewNotification(newNotification) {
```
Please name functions according to our [style guidelines](https://github.com/wlanslovenija/PiplMesh/pull/199#commitcomment-1707793).

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
+function AddNewNotification(newNotification) {
+    $('.notifications').html(parseInt($('.notifications').html())+1);
+    var notif = buildNotification(newNotification.notifications)
+    var content = '<li class="notification">' + notif.author + notif.message + notif.date + '</li>';
```
Don't create HTML fragments manually, use jQuery. Check how we are doing it for posts.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
 }
 
+function AddNewNotification(newNotification) {
+    $('.notifications').html(parseInt($('.notifications').html())+1);
```
You probably want `.text()` instead of `.html()`.

Space around `+`.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
 
+function AddNewNotification(newNotification) {
+    $('.notifications').html(parseInt($('.notifications').html())+1);
+    var notif = buildNotification(newNotification.notifications)
```
Style!

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
+    $('.notifications').html(parseInt($('.notifications').html())+1);
+    var notif = buildNotification(newNotification.notifications)
+    var content = '<li class="notification">' + notif.author + notif.message + notif.date + '</li>';
+    $('.notification_list').prepend(content)
```
`;` at the end.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
+}
+
+function buildNotification(notification) {
+    var new_notif = {};
```
Be verbose! Write it out!

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
+    var new_notif = {};
+    var author = gettext("commented on");
+    var post = gettext("post");
+    new_notif.author = notification.author + ' ' + author + ' <a href="#" >' + post + '</a>.<br />';
```
O no no. You will not translate strings like this. Learn how to interpolate strings with translation in JavaScript in Django.

Don't do `<a href="#" >` links because they might destroy information we might have stored in anchor/hash.

mitar replied:
`<br />` should be unnecessary. Style it with CSS. Use block content? Use `<ul>`/`<li>`?

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
+    var author = gettext("commented on");
+    var post = gettext("post");
+    new_notif.author = notification.author + ' ' + author + ' <a href="#" >' + post + '</a>.<br />';
+    new_notif.message = '<span class="notification_message">' + notification.content + '</span><br />';
```
Same.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
+    var post = gettext("post");
+    new_notif.author = notification.author + ' ' + author + ' <a href="#" >' + post + '</a>.<br />';
+    new_notif.message = '<span class="notification_message">' + notification.content + '</span><br />';
+    new_notif.date = formatDiffTime(notification.created_time);
```
If you use `formatDiffTime` you also have to register automatic updating of diff time as time passes. See how it is done for posts.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
+}
+
+function loadNotifications() {
+    $.getJSON(URLS['notifications'], function (notifications) {
```
`URLS['notifications']` -> `URLS.notifications`.

For constant strings you can access like this.

mitar replied:
Add all arguments callback function receives.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
+        var list = [];
+        var unread_counter = 0;
+        $.each(notifications.objects, function (i, notification) {
+            if (notification.read == false) {
```
`(!notification.read)`?

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
+            if (notification.read == false) {
+                unread_counter += 1;
+            }
+            var notif = buildNotification(notification);
```
Be verbose.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
+                unread_counter += 1;
+            }
+            var notif = buildNotification(notification);
+            list.unshift('<li class="notification">' + notif.author + notif.message + notif.date + '</li>');
```
Don't create fragments like this.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
+            list.unshift('<li class="notification">' + notif.author + notif.message + notif.date + '</li>');
+        })
+
+        var content = '<ul class="notification_list">' + list.join('') + '</ul>';
```
Aaagh. Why you simply don't use jQuery to attach directly to DOM? Why play so much with strings? You would not even need temporary array.

mitar commented on piplmesh/frontend/templates/header.html:
```diff
@@ -31,6 +31,17 @@
                 </li>
                 <li class="username"><a href="{% url "profile" username=user.username %}">{{ user.username }}</a></li>
+                 <li>
+                    <div class="notifications">
```
This would probably better and `id`? There will be only one on the page. And `notifications_count` for example?

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
+        success: function () {
+            alert("Komentar napisan.");
+        },
+        error: function (error) {
```
No need for this, this should already be done by global callback.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
+        url: '/api/v1/post/5050512b6c20b1028c45cf86/scomments/',
+        data: JSON.stringify({'message': comment}),
+        contentType: 'application/json',
+        dataType: "json",
```
Constant string.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,58 @@ function showLastPosts(offset) {
+        data: JSON.stringify({'message': comment}),
+        contentType: 'application/json',
+        dataType: "json",
+        success: function () {
```
Add all arguments to the callback.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -216,6 +268,21 @@ $(document).ready(function () {
     });
 
+    // Notifications
+    $('.notifications').click(function (){
```
Add all arguments to the callback.

Space after `()`.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -216,6 +268,21 @@ $(document).ready(function () {
+    $('.notifications').click(function (){
+        $('#notif_box').slideToggle('fast');
+    });
+    $(".close_notif_box").click(function (){
```
Space after `()`.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -216,6 +268,21 @@ $(document).ready(function () {
+    $(".close_notif_box").click(function (){
+        $('#notif_box').slideToggle('fast');
+    });
+    $('#addCom').click(function (){
```
Same.

mitar commented:
> I am working on page that shows all notification.

Please don't. Please finalize this pull request and open a ticket for this. This can be another pull request, maybe even done by somebody else. I can help you there. But let us finish this here first.

zupan committed:
Changed html parsing, translation in js fixed code style...

zupan committed:
Fixed authorization.py and resources.py.

zupan committed:
Fixed all known issues/errors.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +30,41 @@ def hydrate(self, bundle):
+
+    class Meta:
+        queryset = api_models.Notification.objects.all()
+        allowed_methods = ('get', 'patch')
```
PATCH is still to broad. It means you can update any field in the object. We probably do not want to allow this? You should define all other fields as readonly.

zupan replied:
I already fixed this just didn't pushed it yet. I decided to remove `PATCH` because updating read field will be done in new pull request. I have problem with serializing notification using tastypie as i mentioned in another comment. I wrote all details to you in a mail because this hidden comments are hard to find.

mitar replied:
Good. Then clean this.

Yes, I know. It is not easy problem. So I am leaving it a bit to "steam".

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +30,41 @@ def hydrate(self, bundle):
+        for subscriber in self.instance.subscribers:
+            if subscriber != bundle.obj.author:
+                notification = api_models.Notification.objects.create(recipient=subscriber, post=self.instance, comment=bundle.obj.pk)
+                signals.notification_created.send(sender=self, notification=notification, request=request or bundle.request, bundle=bundle)
```
Yes, then bundle is not reasonable to have here, if it is not notification bundle. Remove it.

mitar commented on piplmesh/api/resources.py:
```diff
@@ -26,12 +30,41 @@ def hydrate(self, bundle):
         # TODO: Make proper authorization, current implementation is for development use only
         authorization = tastypie_authorization.Authorization()
 
+class NotificationResource(resources.MongoEngineResource):
```
Configure other fields read-only.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,63 @@ function showLastPosts(offset) {
+function addComment(comment) {
+    $.ajax({
+        type: 'POST',
+        url: '/api/v1/post/5050512b6c20b1028c45cf86/comments/',
```
TODO.

mitar commented on piplmesh/frontend/views.py:
```diff
@@ -118,6 +118,24 @@ def send_update_on_new_post(sender, post, request, bundle, **kwargs):
+    """
+    serialized = sender.serialize(request, {
+        'type': 'notification',
+        'notification': {'comment_author' : notification.post.comments[int(notification.comment)].author,
```
Rewrap. Remove unnecessary spaces.

mitar commented on piplmesh/frontend/views.py:
```diff
@@ -118,6 +118,24 @@ def send_update_on_new_post(sender, post, request, bundle, **kwargs):
+                        'read': notification.read,
+                   },
+    }, 'application/json')
+    updates.send_update(notification.recipient.get_user_channel(), serialized, True)
```
Why do you serialize yourself? You could leave to `send_update` to serialize for you. In this case you should also not really use `sender.serialize` as it is from wrong class.

mitar commented on piplmesh/frontend/views.py:
```diff
@@ -118,6 +118,24 @@ def send_update_on_new_post(sender, post, request, bundle, **kwargs):
+        'notification': {'comment_author' : notification.post.comments[int(notification.comment)].author,
+                        'recipient': notification.recipient.username,
+                        'comment': int(notification.comment),
+                        'created_time': notification.created_time.isoformat(),
```
Ehm, not good. You should not use ISO format. We are using another format for formatting dates. But `sender.serialize` should take care of this automatically? So maybe leave `sender.serialize` for now, but remove this isoformat.

mitar commented on piplmesh/frontend/templates/notifications.html:
```diff
@@ -0,0 +1,9 @@
+{% extends "plain.html" %}
```
This file is probably unnecessary in this pull request?

mitar commented on piplmesh/frontend/tasks.py:
```diff
@@ -46,6 +46,9 @@ def check_online_users():
             connection_last_unsubscribe__lt=timezone.now() - datetime.timedelta(seconds=CHECK_ONLINE_USERS_RECONNECT_TIMEOUT),
         ).update(set__is_online=False):
+            user.reload()
+            user.channel_id = models.generate_channel_id()
```
Some comment would be really good here, why and what exactly we are doing here. So why exactly here, what is this.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -216,6 +273,21 @@ $(document).ready(function () {
+    $('.close_notifications_box').click(function () {
+        $('#notifications_box').slideToggle('fast');
+    });
+    $('#add_comment').click(function () {
```
Click callback gets `event` argument.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -216,6 +273,21 @@ $(document).ready(function () {
+    $('#notifications_count').click(function () {
+        $('#notifications_box').slideToggle('fast');
+    });
+    $('.close_notifications_box').click(function () {
```
Same.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -216,6 +273,21 @@ $(document).ready(function () {
     });
 
+    // Notifications
+    $('#notifications_count').click(function () {
```
Same.

mitar replied:
TODO.

mitar replied:
You haven't added `event` here.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,63 @@ function showLastPosts(offset) {
+        })
+
+        $('#notifications_content').html(content);
+        $('#notifications_count').html(unread_counter);
```
`.text`, `unread_counter` does not contain any HTML, no?

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,63 @@ function showLastPosts(offset) {
+
+        $.each(notifications.objects, function (i, notification) {
+            if (!notification.read) {
+                unread_counter += 1;
```
JavaScript has `unread_counter++`.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,63 @@ function showLastPosts(offset) {
 
+function addNewNotification(newNotification) {
+    var notification_counter = parseInt($('#notifications_count').text()) + 1;
+    $('#notifications_count').html(notification_counter);
```
`.text`.

zupan replied:
Fixed all that was mentioned above. I thing everything is set for pull.

mitar replied:
You have not changed to `.text` here.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,63 @@ function showLastPosts(offset) {
+    ).append(
+        $('<span/>').addClass('notification_element').addClass('notification_created_time').text(formatDiffTime(notification.created_time))
+    );
+    new_notification.data('data', notification);
```
Ehm, better name for this data could be useful. What posts have? `post?` Then `notification` here?

mitar commented on piplmesh/account/models.py:
```diff
@@ -1,6 +1,6 @@
 from __future__ import absolute_import
 
-import datetime, hashlib, urllib
+import datetime, hashlib, os, urllib
```
`os` is not needed anymore?

mitar commented on piplmesh/account/models.py:
```diff
@@ -8,12 +8,14 @@
 from django.db import models
 from django.test import client
-from django.utils import timezone
+from django.utils import crypto, timezone
```
`crypto` too?

zupan committed:
Fixed js in home.js added some comments and TODO.

mitar commented on piplmesh/api/signals.py:
```diff
@@ -1,3 +1,4 @@
 from django import dispatch
 
 post_created = dispatch.Signal(providing_args=('post', 'request', 'bundle'))
+notification_created = dispatch.Signal(providing_args=('notification', 'request', 'bundle'))
```
Remove `bundle` here.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,65 @@ function showLastPosts(offset) {
+function addComment(comment) {
+    $.ajax({
+        type: 'POST',
+        // TODO: This url has to be dynamic
```
No, you should make pull request to work as it is. For things which are not yet in the pull request, you can add TODO or new ticket. But for things which are in the pull request, you should make it so that after it is pulled, things work. And this does not work for nobody else, except if by luck it has post with exactly this ID.

mitar commented on piplmesh/frontend/views.py:
```diff
@@ -118,6 +118,25 @@ def send_update_on_new_post(sender, post, request, bundle, **kwargs):
+    """
+    Sends update to push server when a new notification is created.
+    """
+    serialized = sender.serialize(request, {
```
Add TODO.

zupan committed:
Changed addComment and fixed some issues with variable names.

zupan committed:
Added TODO in frontend/views.py.

zupan committed:
Updated translations.

mitar commented on piplmesh/frontend/static/piplmesh/js/home.js:
```diff
@@ -197,6 +197,67 @@ function showLastPosts(offset) {
+
+    $.ajax({
+        type: 'POST',
+        url: post_url + 'comments/',
```
Mnja. Let it be. :-) But you should add TODO, that it should not be hardcoded like this. :-)

mitar commented on piplmesh/locale/sl/LC_MESSAGES/djangojs.po:
```diff
@@ -72,3 +75,9 @@ msgstr "Slika uporabnika"
 msgid "User profile"
 msgstr "Profil uporabnika"
+
+#~ msgid "commented on"
```
Unused strings should be cleared. Especially if they have been made by you during developing of this pull request.

mitar commented:
I have updated and cleaned the code and fixed serialization. I opened two more tickets: #300 and #301. Was there anything else we talked that it is needed to be done?

Feel free to continue working on notifications now.

mitar commented:
(And please check changes I have done and try to understand them and also reasons for them.)
